databuilder/query_language.py

```diff
@@ -147,7 +147,7 @@ def sort_by(self, series):
         )
 
 
-class SortedEventFrame(Frame):
+class SortedEventFrame(EventFrame):
```

Comment:
I don't think this will quite work as you intend. It enables `e.sort_by().sort_by().first_for_patient()` (which you are quite right that we do need), but `e.sort_by().take().first_for_patient()` because `take()` returns an `EventFrame` which has no `first_for_patient()` method.

It would be _reasonable_, I think, to insist that filtering has to happen before sorting in order to enforce uniformity (I think this was the original intention) -- but there's also an argument for maximizing flexibility.

Maybe you could add specs to cover this behaviour to completely pin it down?

Reply:
Good point. Although this change does make ehrQL exactly reflect the semantics of the QM, so in that sense it does achieve what I intended. But the semantics of the QM are weird in that it allows any mixture of "sorts" and "takes" but require that at least one of the "sorts" immediately precedes the final "pick", and I hadn't thought about that enough.

Assuming we can't live with that behaviour I think we have two options. One is to track sortedness dynamically, as a property that is passed through Frame operations, just as we track domains through Series operations. The difficulty here isn't so much the tracking as it is finding a readable and obvious way of expressing the rules as part of the declaration of the QM classes. At the moment we _mostly_ manage to do this with type-annotations, apart from the domain constraints. (I was happier when there was exactly one domain rule that applied everywhere here, but it's not too bad with just the Sort and Filter exceptions.)

This is going to be tricky, but probably we're going to have to handle this anyway because we're going to need to find some way of preventing aggregations being run on patient-series and that's a structurally parallel problem.

Another option is that we stop requiring sorting at all before a pick operation. Given that we're going to need to impose an [implicit sort order](https://github.com/opensafely-core/databuilder/issues/461) anyway then this is still a well-defined operation. And it would obviously be the technically easiest solution. My worry is that it would be easy for users to assume that `first_by_patient()` came with an implicit date ordering if they're allowed to use it without sorting first. But maybe we should allow it in the QM and then have some extra anti-footgun validation in ehrQL?

Finally, I think we're going to need some way of expanding our auto-documenting spec tests to include specifying what is and isn't valid ehrQL. At the moment, I don't think there's any way of doing this.

Reply:
Happy for you to take whatever path you like through this thicket.